cnxarchive/views/content.py

```diff
@@ -0,0 +1,221 @@
+                            filename='index.cnxml.html')
+                cursor.execute(SQL['get-resource-by-filename'], args)
+                try:
+                    content = cursor.fetchone()[0]
+                except (TypeError, IndexError,):  # None returned
+                    logger.debug("module found, but "
+                                 "'index.cnxml.html' is missing.")
+                    raise httpexceptions.HTTPNotFound()
+                result['content'] = content[:]
+
+    return result
+
+
+def html_listify(tree, root_ul_element, parent_id=None):
+    """Recursively construct HTML nested list version of book tree.
+    The original caller should not call this function with the
+    `parent_id` defined.
+
+    """
+    request = get_current_request()
+    is_first_node = parent_id is None
+    if is_first_node:
+        parent_id = tree[0]['id']
+    for node in tree:
+        li_elm = etree.SubElement(root_ul_element, 'li')
+        a_elm = etree.SubElement(li_elm, 'a')
+        a_elm.text = node['title']
+        if node['id'] != 'subcol':
+            if is_first_node:
+                a_elm.set('href', request.route_path(
+                    'content-html', ident_hash=node['id']))
+            else:
+                a_elm.set('href', request.route_path(
+                    'content-html',
+                    separator=':',
+                    ident_hash=parent_id,
+                    page_ident_hash=node['id']))
+        if 'contents' in node:
+            elm = etree.SubElement(li_elm, 'ul')
+            html_listify(node['contents'], elm, parent_id)
+
+
+# ######### #
+#   Views   #
+# ######### #
+
+
+@view_config(route_name='content-json', request_method='GET')
+def get_content_json(request):
+    """Retrieve content as JSON using the ident-hash (uuid@version)."""
+    result = _get_content_json()
+
+    result = json.dumps(result)
+    resp = request.response
+    resp.status = "200 OK"
+    resp.content_type = 'application/json'
+    resp.body = result
+    return resp
+
+
+@view_config(route_name='content-html', request_method='GET')
+def get_content_html(request):
+    """Retrieve content as HTML using the ident-hash (uuid@version)."""
+    result = _get_content_json()
+
+    media_type = result['mediaType']
+    if media_type == COLLECTION_MIMETYPE:
+        content = tree_to_html(result['tree'])
+    else:
+        content = result['content']
+
+    resp = request.response
+    resp.status = "200 OK"
+    resp.content_type = 'application/xhtml+xml'
+    resp.body = content
+    return resp
+
+
+@view_config(route_name='content', request_method='GET')
+def get_content(request):
+    """Retrieve content using the ident-hash (uuid@version).
+
+    Depending on the HTTP_ACCEPT header return HTML or JSON.
+    """
+    if 'application/xhtml+xml' in request.headers.get('ACCEPT', ''):
+        return get_content_html(request)
+
+    else:
+        return get_content_json(request)
```

Comment:
Ah, but its easier to find since the python filename matches the route. Which is does, now.